bee-network/src/config.rs

```diff
@@ -0,0 +1,402 @@
+    pub fn replace_addr(&mut self, mut addr: Protocol) {
+        if !matches!(addr, Protocol::Ip4(_) | Protocol::Ip6(_) | Protocol::Dns(_)) {
+            panic!("Invalid addr");
+        }
+
+        if let Protocol::Dns(dns) = addr {
+            addr = resolve_dns_multiaddr(dns);
+        }
+
+        let port = self.bind_multiaddr.pop().expect("multiaddr pop");
+        let _ = self.bind_multiaddr.pop().expect("multiaddr pop");
+        self.bind_multiaddr.push(addr);
+        self.bind_multiaddr.push(port);
+    }
+
+    /// Replaces the port of the bind address.
+    ///
+    /// The argument `port` must be the TCP variant of [`Protocol`].
+    pub fn replace_port(&mut self, port: Protocol) {
+        if !matches!(port, Protocol::Tcp(_)) {
+            panic!("Invalid port");
+        }
+
+        self.bind_multiaddr.pop();
+        self.bind_multiaddr.push(port);
+    }
+
+    /// Adds a static peer.
+    pub fn add_static_peer(&mut self, peer_id: PeerId, multiaddr: Multiaddr, alias: Option<String>) {
+        self.static_peers.push(Peer {
+            peer_id,
+            multiaddr,
+            alias,
+        });
+    }
+}
+
+fn resolve_dns_multiaddr(dns: Cow<'_, str>) -> Protocol {
+    use std::net::{IpAddr, ToSocketAddrs};
+
+    for socket_addr in dns.to_socket_addrs().expect("Invalid Multiaddr: Unresolvable") {
+        match socket_addr.ip() {
+            IpAddr::V4(ip4) => return Protocol::Ip4(ip4),
+            IpAddr::V6(ip6) => return Protocol::Ip6(ip6),
+        }
+    }
+    panic!("Invalid Multiaddr: Unresolvable");
+}
+
+impl Default for NetworkConfig {
+    fn default() -> Self {
+        Self {
+            // Panic:
+            // Unwrapping is fine, because we made sure that the default is parseable.
+            bind_multiaddr: DEFAULT_BIND_MULTIADDR.parse().unwrap(),
+            reconnect_interval_secs: DEFAULT_RECONNECT_INTERVAL_SECS,
+            max_unknown_peers: DEFAULT_MAX_UNKOWN_PEERS,
+            static_peers: Vec::new(),
+        }
+    }
+}
+
+/// A network configuration builder.
+#[derive(Default, Deserialize)]
+pub struct NetworkConfigBuilder {
+    #[serde(rename = "bind_address")]
+    bind_multiaddr: Option<Multiaddr>,
+    reconnect_interval_secs: Option<u64>,
+    max_unknown_peers: Option<usize>,
+    peering: PeeringConfigBuilder,
+}
+
+impl NetworkConfigBuilder {
+    /// Creates a new default builder.
+    pub fn new() -> Self {
+        Self::default()
+    }
+
+    /// Specifies the bind addresses.
+    pub fn with_bind_multiaddr(mut self, mut multiaddr: Multiaddr) -> Self {
+        let mut valid = false;
+        let mut is_dns = false;
+
+        for (i, p) in multiaddr.iter().enumerate() {
+            match i {
+                0 => {
+                    if !matches!(p, Protocol::Ip4(_) | Protocol::Ip6(_) | Protocol::Dns(_)) {
+                        panic!("Invalid Multiaddr: at index 0")
```

Comment:
Should this function be returning an error instead?

Reply:
there is no reason to return an error (I think) because if the configuration of the network is faulty, I want to panic.

Reply:
Is that not a choice to be made by the caller though given that this is a public API?

Reply:
I guess you're right. Changed it!